daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/Util.scala

```diff
@@ -84,4 +84,69 @@ object Util {
   val CPTrue = CPPrimCon(PCTrue)
   val CPFalse = CPPrimCon(PCFalse)
 
+  type Graph[X] = Map[X, Set[X]]
+
+  // Topologically order the vertices of an abstract Graph.
+  // If the `graph` is a directed acyclic graph returns a list of its vertices in topological order as `Right`
+  // otherwise returns a list of vertices that form a cycle as `Left`
```

Comment:
If you wrapped that list of vertices that form a cycle in something like `case class Cycle`, it'd be harder to do the wrong thing with it.